packages/beacon-node/src/chain/validation/attestation.ts

```diff
@@ -568,41 +572,116 @@ export function verifyHeadBlockAndTargetRoot(
+ *
+ * This implementation does not require to dial head state to attSlot at fork boundary because we always get domain of attSlot
+ * in consumer context.
  *
- * Otherwise, regenerate state from head state dialing to target epoch
+ * This is similar to the old getStateForAttestationVerification
+ * see https://github.com/ChainSafe/lodestar/blob/v1.11.3/packages/beacon-node/src/chain/validation/attestation.ts#L566
  */
-export async function getStateForAttestationVerification(
+export async function getShufflingForAttestationVerification(
   chain: IBeaconChain,
-  attSlot: Slot,
   attEpoch: Epoch,
   attHeadBlock: ProtoBlock,
   regenCaller: RegenCaller
-): Promise<CachedBeaconStateAllForks> {
-  const isSameFork = chain.config.getForkSeq(attSlot) === chain.config.getForkSeq(attHeadBlock.slot);
-  // thanks for 1 epoch look ahead of shuffling, a state at epoch n can get committee for epoch n+1
-  const headStateHasTargetEpochCommmittee =
-    attEpoch - computeEpochAtSlot(attHeadBlock.slot) <= SHUFFLING_LOOK_AHEAD_EPOCHS;
+): Promise<EpochShuffling | null> {
+  const blockEpoch = computeEpochAtSlot(attHeadBlock.slot);
+  let shufflingDependentRoot: RootHex;
+  if (blockEpoch === attEpoch) {
+    // current shuffling, this is equivalent to `headState.currentShuffling`
+    // given blockEpoch = attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     ------------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.previous);
+  } else if (blockEpoch === attEpoch - 1) {
+    // next shuffling, this is equivalent to `headState.nextShuffling`
+    // given blockEpoch = n-1, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.current);
+  } else if (blockEpoch < attEpoch - 1) {
+    // this never happens with default chain option of maxSkipSlots = 32, however we still need to handle it
+    // check the verifyHeadBlockAndTargetRoot() function above
+    // given blockEpoch = n-2, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -----------^
+    // shufflingDependentRoot -----^
+    shufflingDependentRoot = attHeadBlock.blockRoot;
+    // use lodestar_gossip_attestation_head_slot_to_attestation_slot metric to track this case
+  } else {
+    // blockEpoch > attEpoch
+    // should not happen, handled in verifyAttestationTargetRoot
+    throw Error(`attestation epoch ${attEpoch} is before head block epoch ${blockEpoch}`);
+  }
```

Comment:
Move all of this to a separate function, like getShufflingDependentRoot. What you have an if else assignment pattern, it's always a good indication to move the code to a function to use the return flow control

Reply:
addressed in #6030 👍